Add inductor OrderedSet

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130004
* __->__ #130003

Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Add inductor OrderedSet

[ghstack-poisoned]

aorenste commented:
Should this be in _inductor or in torch.utils? Seems like we might want to use it outside inductor.

aorenste commented on torch/_inductor/ordered_set.py:
```diff
@@ -0,0 +1,129 @@
+    def difference(self, *others: Iterable[T]) -> "OrderedSet[T]":
+        res = self if len(others) else self.copy()
+        for other in others:
+            res = res - other  # type: ignore[operator, arg-type]
```
This `type: ignore` seems suspect - I think the error is actually correct:
```
(base) $ python -c 'print(set((1, 2, 3)) - [2])'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: unsupported operand type(s) for -: 'set' and 'list'```
```
vs
```
(base) $ python -c 'print(set((1, 2, 3)).difference([2]))'
{1, 3}
```
So it seems that '-' actually doesn't support generic Iterables whereas difference does.  I guess they need _wrap_in_set as well.

(I assume the same thing applies to the following functions w/ `type: ignore` as well)

eellison replied:
I'm inheriting from collections.ABC.mutableset which allows this behavior. I mentioned this in the summary. See, this code here:

https://github.com/python/cpython/blob/3.12/Lib/_collections_abc.py#L642-L644

I would rather inherit from there behavior although we could decide to throw if want.

aorenste replied:
This is fine - I didn't realize that MutableSet provided this functionality.  It's kind of weird that they support iterable but their signature takes AbstractSet.

ezyang commented:
There's no reason to put this in torch._inductor. Put it in utils please.

Skylion007 commented on torch/_inductor/ordered_set.py:
```diff
@@ -0,0 +1,129 @@
+from collections.abc import MutableSet, Set as AbstractSet
+from typing import Any, cast, Iterable, Iterator, Optional, TypeVar
+
+T = TypeVar("T")
```
nit: private TypeVar should probably start with underscore.

aorenste commented on torch/_inductor/ordered_set.py:
```diff
@@ -0,0 +1,129 @@
+        return elem in self._dict
+
+    def __iter__(self) -> Iterator[T]:
+        used = len(self)
+        it = iter(self._dict)
+        for _ in range(used):
+            if len(self) != used:
+                raise RuntimeError("OrderedSet changed size during iteration")
+            yield next(it)
```
(I guess in this case it could also be `return self._dict.keys()`)

aorenste commented on torch/_inductor/ordered_set.py:
```diff
@@ -0,0 +1,129 @@
+    # Required overriden abstract methods
+    #
+
+    def __contains__(self, elem: Any) -> bool:
```
> You're probably right but there are 930 uses of `: Any` in torch and 141 uses of `: object`. I don't like really forking the idiom in two :/
> 
> Will update tho

Just because there are 930 incorrect uses of `: Any` doesn't make it "correct"...

To be clear - we're not just talking about two names for the same thing - `Any` and `object` are different and use of `Any` is almost always wrong (because it turns off typechecking). In this case it's moot because we're passing an `Any` into an `object` on an interface that can never change so the end effect is the same - but it's still "wrong" because if the thing we were passing it into were made more specific (it could never happen but imagine if python changed `__contains__` to `elem: T`) then mypy would be no help in finding those mistakes.

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste reviewed (approved):
LGTM - Less of a big deal but should the test file be moved as well?  I guess we don't have a `test/utils/` directory.

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

XuehaiPan commented on torch/utils/ordered_set.py:
```diff
@@ -0,0 +1,141 @@
+    __slots__ = ("_dict",)
+
+    def __init__(self, iterable: Optional[Iterable[T]] = None):
+        self._dict: dict[T, None] = {}
+        if iterable is not None:
+            self._dict = {k: None for k in iterable}
```
```suggestion
            self._dict = dict.fromkeys(iterable)
```

XuehaiPan replied:
```suggestion
        self._dict: dict[T, None] = OrderedDict()
        if iterable is not None:
            self._dict = OrderedDict.fromkeys(iterable)
```

XuehaiPan commented on torch/utils/ordered_set.py:
```diff
@@ -0,0 +1,141 @@
+        return list(self._dict.keys())
+
+    def __setstate__(self, state: List[T]) -> None:
+        self._dict = {k: None for k in state}
```
```suggestion
        self._dict = dict.fromkeys(state)
```

XuehaiPan replied:
```suggestion
        self._dict = OrderedDict.fromkeys(iterable)
```

XuehaiPan commented on torch/utils/ordered_set.py:
Maybe we should implement a `__eq__` method like `OrderedDict` does: compare the key order if `other` is also an `OrderedDict`.

```python
In [1]: from collections import OrderedDict

In [2]: {'a': 1, 'b': 2} == {'b': 2, 'a': 1}
Out[2]: True

In [3]: OrderedDict({'a': 1, 'b': 2}) == {'b': 2, 'a': 1}
Out[3]: True

In [4]: {'a': 1, 'b': 2} == OrderedDict({'b': 2, 'a': 1})
Out[4]: True

In [5]: OrderedDict({'a': 1, 'b': 2}) == OrderedDict({'b': 2, 'a': 1})
Out[5]: False
```

```python
def __eq__(self, other):
    if isinstance(other, OrderedSet):
        return list(self) == list(other)
    return super().__eq__(other)
```

XuehaiPan commented on torch/utils/ordered_set.py:
```diff
@@ -0,0 +1,141 @@
+        # overridden because MutableSet impl is slow
+        self._dict.clear()
+
+    # Unimplemented set() methods in _collections_abc.MutableSet
```
Implement `OrderedSet.move_to_end()` from `OrderedDict.move_to_end()`

```suggestion
    # Unimplemented set() methods in _collections_abc.MutableSet
    
    def move_to_end(self, value: T, last: bool = True) -> None:
        self._dict.move_to_end(value, last=last)
```

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

XuehaiPan commented on torch/utils/ordered_set.py:
```diff
@@ -0,0 +1,174 @@
+            self._dict = dict.fromkeys(iterable, None)
+
+    @staticmethod
+    def from_dict(dict_inp: Dict[T, None]) -> OrderedSet[T]:
```
```suggestion
    def _from_dict(dict_inp: Dict[T, None]) -> OrderedSet[T]:
```

It should be an implementation detail and be private. The public "from_dict" semantic should be:

```python
OrderdedSet(dict_)
```

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Add inductor OrderedSet"


Implemented by extending `collections.abc.MutableSet` and backing it with a dictionary, which is ordered. From collections.abc.MutableSet:

```
    A mutable set is a finite, iterable container.

    This class provides concrete generic implementations of all
    methods except for __contains__, __iter__, __len__,
    add(), and discard().
```

In addition to implementing those methods I also had to define some methods of python's set which were not implemented in MutableSet.

I reused the test from my python's lib. There were a few instances of tests that didnt pass because edge case behavior that is not necessary to reimplement
- support self-referencing repr
- erroring when an member's `__eq__` function would modify the set itself
- MutableSet supports Iterables as inputs, but not sequences (pretty rare..)
- Some specifics of exact equivalent type errors being thrown
- [The protocol for automatic conversion to immutable](https://docs.python.org/2/library/sets.html#protocol-for-automatic-conversion-to-immutable)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

XuehaiPan commented on torch/utils/_ordered_set.py:
```diff
@@ -0,0 +1,179 @@
+
+    __slots__ = ("_dict",)
+
+    def __init__(self, iterable: Optional[Iterable[T]] = None):
+        self._dict = dict.fromkeys(iterable, None) if iterable is not None else {}
```
```suggestion
    def __init__(self, iterable: Iterable[T] = ()):
        self._dict = dict.fromkeys(iterable, None)
```

eellison replied:
will do in follow up

eellison commented:
@pytorchbot merge -f "large amount of code is from python tests"
